_source/_posts/2022-10-01-spring-vault.md

```diff
@@ -0,0 +1,639 @@
+--name my-vault vault
+```
+
+**NOTE**:  The `docker run` command above will start a vault instance with the name `my-vault`. You can stop the container with `docker stop my-vault` and restart it with `docker start my-vault`. Note that all the secrets and data will be lost between restarts, as explained in the next paragraphs.
+
+IPC_LOCK capability is required for Vault to be able to lock memory and not be swapped to disk, as this behavior is enabled by default. As the instance is run for development, the ID of the initially generated root token is set to the given value. We are mounting `/vault/logs`, as we are going to enable the `file` audit device to inspect the interactions.
+
+Once it starts, you should notice the following logs:
+
+```
+WARNING! dev mode is enabled! In this mode, Vault runs entirely in-memory
+and starts unsealed with a single unseal key. The root token is already
+authenticated to the CLI, so you can immediately begin using Vault.
+
+You may need to set the following environment variable:
+
+    $ export VAULT_ADDR='http://0.0.0.0:8200'
+
+The unseal key and root token are displayed below in case you want to
+seal/unseal the Vault or re-authenticate.
+
+Unseal Key: +xTAEzkjb3kurSBa7TxAQUjoyBpaprlXwn+ZnSpEkuw=
+Root Token: 00000000-0000-0000-0000-000000000000
+
+Development mode should NOT be used in production installations!
+```
+
+It is clear Vault is running in _dev mode_, meaning it short-circuits a lot of setup to insecure defaults, which helps for the experimentation. Data is stored encrypted in-memory and lost on every restart. Copy the _Unseal Key_, as we are going to use it to test Vault sealing.
+Connect to the container and explore some vault commands:
+
+```shell
+docker exec -it my-vault /bin/sh
+```
+
+The command above will start a shell session with the container. After the prompt shows, run the following three commands:
+
+```shell
+export VAULT_TOKEN="00000000-0000-0000-0000-000000000000"
+export VAULT_ADDR="http://127.0.0.1:8200"
+vault status
+```
+
+The `status` command output shows if the vault instance is sealed:
+
+```
+Key             Value
+---             -----
+Seal Type       shamir
+Initialized     true
+Sealed          false
+Total Shares    1
+Threshold       1
+Version         1.11.3
+Build Date      2022-08-26T10:27:10Z
+Storage Type    inmem
+Cluster Name    vault-cluster-389dcb61
+Cluster ID      cc063fcf-e1f2-2e75-6059-cd9117559f0d
+HA Enabled      false
+```
+
+As you can see, in development mode Vault starts unsealed, meaning stored data can be decrypted/accessed.
+
+Enable a file audit device to watch the interactions with Vault:
+
+```shell
+vault audit enable file file_path=/vault/logs/vault_audit.log
+```
+
+You should see a success message. Now store the Okta secrets for the `vault-demo-app`:
+
+```shell
+vault kv put secret/vault-demo-app,dev \
+okta.oauth2.clientId="{yourClientId}" \
+okta.oauth2.clientSecret="{yourClientSecret}" \
+okta.oauth2.issuer="{yourIssuerURI}"
```

Comment:
At this point, these values are gone because I replaced them in the vault-demo-app. Maybe we should add a sentence to save them for later or recommend `okta apps config --app clientId `?

Reply:
Added a comment.